Implemented retry on error

moser4035 opened the pull request.

Commit:
Implemented executeHttpRequestAndRetryOnError() and changed to retry if there is an error.
Changed to not use lamda in getMuxings() to obtain java 1.7 compatibility
Changed assertion in testAddAndDeleteDomain() to use CollectionUtils.isNotEmpty() instead of is() matcher (Java 1.7 compatibility)
Adjusted RestClient to set retry
Adjusted BItmovinApi to set retry
Added tests

msmole commented on src/main/java/com/bitmovin/api/RestClient.java:
```diff
@@ -398,6 +399,15 @@ public static Long getTotalCount(String url, Map<String, String> headers) throws
+
+    public static void setRetry(boolean retry)
+    {
+        RestClient.retry = retry;
```
You also need to set `this.retry = retry`?

moser4035 replied:
This is a static field, so this is set that way

msmole commented on src/main/java/com/bitmovin/api/http/JsonRestClient.java:
```diff
@@ -200,9 +203,77 @@ public void delete(URI resource, Map<String, String> headers, Object content) th
+        {
+            return this.executeHttpRequestWithRetries(httpRequest, responseClass);
+        }
+        else
```
No else needed because of previous early return.

Commit:
Merge branch 'develop' into feature/rest_client_retry_on_error

Commit:
Removed unnecessary else

Commit:
Implemented to only retry on 500 errors
Adjusted tests to check if retry is working correctly and does not retry on other

jamdrop commented on src/main/java/com/bitmovin/api/http/JsonRestClient.java:
```diff
@@ -132,77 +135,141 @@ private void checkOrThrowError(HttpResponse response) throws BitmovinApiExceptio
+        for (int i = 1; i <= this.maxRetries; i++)
+        {
+            HttpResponse<String> httpResponse = httpRequest.asString();
+            if (httpResponse.getStatus() >= 500 && httpResponse.getStatus() < 600)
```
LGTM

we get the following error codes:

500:
from the service itself, if it throws an exception

502:
from the google loadbalancer if we have tcp connection issues to envoy (shouldnt happen)

504:
from the google loadbalancer if the envoy, kong, or the service is not available or times out
on GET requests google automatically retries the request and returns with 504 after 60secs

in case of our SQL issue c3p0 does 90 retries each after 1000 millis, thats longer than the http timeout of our loadbalancer, so we are seeing 504 if sql times out

Commit:
Merge branch 'develop' into feature/rest_client_retry_on_error

# Conflicts:
#	pom.xml
#	src/main/java/com/bitmovin/api/resource/encoding/EncodingMuxingResource.java

Commit:
Changed CLOUD_REGION in example to CloudRegion.AUTO

kymr commented on src/main/java/com/bitmovin/api/http/JsonRestClient.java:
```diff
@@ -132,77 +135,141 @@ private void checkOrThrowError(HttpResponse response) throws BitmovinApiExceptio
+        this.maxRetries = maxRetries;
+    }
+
+    public void setMsToWaitForRetrying(int msToWaitForRetrying)
```
What about long instead of int for unit of ms?

kymr commented on src/main/java/com/bitmovin/api/http/JsonRestClient.java:
```diff
@@ -132,77 +135,141 @@ private void checkOrThrowError(HttpResponse response) throws BitmovinApiExceptio
+                continue;
+            }
+
+            return httpRequest.asObject(responseClass);
```
For normal case, redundant request happens.

httpRequest.asString(); // first request executed without error
httpRequest.asObject(responseClass); // second request will be executed

moser4035 replied:
You are right, thanks for your feedback. This is not possible to do this way. I think i have to find another solution for that.
